### ISSUE TITLE
2020 Code Review Party - Team 4183 Feedback

### ISSUE THREAD

**Issue by cpostbitbuckets** [maverick-boiler-robotics-team-4272/2020Robot#9]
# 2020 Code Review Party - Team 4183 Feedback
Hi Team 4272, thank you for sharing your code with our team. Three of our software students and our software mentor all cloned your codebase, reviewed the code, and discussed it as a team to come up with this feedback. We approached the codebase as if we were new students on your team trying to understand how the code works.

## Overall Thoughts
Our first impression of the codebase was that it had an unusual structure compared to other robotics projects we have looked at. Normally we would look for subsystems or a RobotContainer class, but we didn't see any of that. However, once we started looking through the code itself, it wasn't too hard to follow what was going on.

## Things we Liked
- **buttons.txt** - The buttons.txt file is great. Having a single place to look for user interface controls is a great idea and we will be using something similar in the future.
- **Fun Drive Control** - Having the driver operate two joysticks to control the robot and press the fire button to shoot balls sounds super fun.
- **NetworkTables.java** - We did not realize you could setup listeners to listen for NetworkTables changes. We always poll for values during each robotPeriodic. We will definitely be using this in the future.

### Out of Date Libraries
- [ ] Update vendor libraries to the latest version. We were not able to get the project to build until we did this.

### Non-standard File Structure and Organization
The file structure was very strange. We would suggest
- [ ] Organize your code into [subsystems](https://docs.wpilib.org/en/stable/docs/software/commandbased/subsystems.html). This will help new students pick up your code faster and will make it easier for other teams to help you if you post on chiefdelphi.
- [ ] Use java packages for your subsystem to better separate areas of responsibility.

### Comments
- [ ] Add comments above each major class (Shooter, Hopper, HwMotor, etc) summarizing what the class is used for. We were able to figure out most of them, but more complex classes like the Hopper were hard to figure out without reaching out for help.

- [ ] Add comments to the Hopper if/else block in the movement() function. There are so many state changes, it is really hard to figure out what is going on.

### Constants instead of Magic Numbers
- [ ] There were many places in the code with magic numbers. Constants would be more readable and ensure there is only one place for constant values.

For example, this code in Hopper.java:

```java
shooter_infeed.set(-1);
hopper_infeed.set(-0.2);
hopper.set(-0.2);
```

Would be more clear as
```java
// These constants would be declared at the top of Hopper.java, or in a separate HopperConstants.java file

// These constants are the percentage to run the hopper
// subsystem motors at when the shooter is up to speed.
public static final SHOOTER_INFEED_SHOOTING_OUTPUT = -1;
public static final HOPPER_INFEED_SHOOTING_OUTPUT = -.2;
public static final HOPPER_SHOOTING_OUTPUT_PERCENT = -.2;

shooter_infeed.set(SHOOTER_INFEED_SHOOTING_OUTPUT);
hopper_infeed.set(HOPPER_INFEED_SHOOTING_OUTPUT);
hopper.set(HOPPER_SHOOTING_OUTPUT_PERCENT);
```

Another great place for constants in the HwMotor.java class:

```java
    // All motor ids are defined in MotorIds.java
	public final CANSparkMax left1 = new CANSparkMax(LEFT_DRIVE_MOTOR_ID, MotorType.kBrushless);
	public final CANSparkMax left2 = new CANSparkMax(LEFT_DRIVE_MOTOR_FOLLOWER_ID, MotorType.kBrushless);
```

### State Machines
The Hopper logic was very complicated and would probably be better suited for a state machine. Rather than a series of if/else blocks with motor.set() calls, You could create various states the Hopper could be in:
* Feed Balls Through Hopper
* Feed Balls to Shooter
* Don't Feed Balls

The state would be defined as an enum and your logic would set state, i.e.:

```java
...
                if (intake_to_hopper_sensor) {
                    if (!hopper_ball_c) {
                        if (!hopper_ball_b) {
                            if (!hopper_ball_a) {
                                state = HopperState.RunBallsThrough;
                            } else {
                                state = HopperState.RunBallsThrough;
                            }
                        } else {
                                state = HopperState.DoNotRun;
                        }
...

```

Then later on in the code you can use a simple switch statement to command your motors.

```java
switch (state) {
    case HopperState.RunBallsThrough:
        // Run balls through the hopper but not into the shooter
        shooter_infeed.set(0);
        hopper_infeed.set(belt_speed);
        hopper.set(belt_speed);
    break;
    case HopperState.DoNotRun;
        // Don't run anything through the hopper
        shooter_infeed.set(0);
        hopper_infeed.set(0);
        hopper.set(0);
    break;
    ... // Other states
}
```

The advantage of this approach is it keeps your decision making separate from the code to command hardware, it reduces duplicate code, and it makes it clear what conditions lead the hopper to be in each state.

### Minor Things
- [ ] Use a switch statement instead of a bunch of if/else ifs in ColorThing.java
- [ ] Use an enum instead of string literals to choose colors in ColorThing.java. Java doesn't compare strings with ==, you have to use colorString.equals("Yellow"). An enum solves this problem and the compiler will catch any issues with mistyping a string.
- [ ] Use an enum instead of 0, 1, and 2 for NewAuto.java.
- [ ] Clean up some of the indentation in the case statements for NewAuto.java

## Summary
Thank you for giving us the opportunity to see your code. We learned from it and we hope our feedback is helpful to you in your future robot coding. Good luck in the rest of the season!

**Comment by Karaoke4272** [comment]
Thank you for taking the time to review our code! Your review is very informative, and we will definitely take this to heart next time we get the opportunity to program our robot. Good luck to you as well!
